### PR TITLE
binary_to_hex challenge

### PR DESCRIPTION
**Pull request by MarcDarazi99**
I did the challenge called binary_to_hex and it's ready for reviewing

---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [ ] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [ ] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called at the top level of the function file 
  - _Recursive solutions **can** call the function from **inside** the function body_

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [ ] The function's strategy _is not_ described in the documentation
- [ ] Comments explain the _strategy_, **not** the _implementation_
- [ ] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [ ] The code passes the formatting checks
- [ ] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [ ] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [ ] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/binary_to_hex.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+This module provides a utility function to convert binary strings to their 
+corresponding hexadecimal representations. It validates the input to ensure 
+that it contains only binary digits (0s and 1s) and performs the conversion 
+accurately.
+
+Functions:
+    - binary_to_hex(binary_str): Converts a binary string to its hexadecimal representation.
+
+Usage Example:
```
Comment (Moealfadil) [inline solutions/binary_to_hex.py:12]: behavior of function should be discussed within the docstring of the function, put all examples in the docstring below

File: solutions/tests/test_binary_to_hex.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    - binary_to_hex(binary_str): Converts a binary string to its hexadecimal representation.
+
+Usage Example:
+    >>> from binary_to_hex import binary_to_hex
+    >>> binary_to_hex("1101")
+    'D'
+
+"""
+
+def binary_to_hex(binary_str):
```
Comment (Moealfadil) [inline solutions/tests/test_binary_to_hex.py:19]: the function should be written inside the test file. import the function instead using "from (filename) import (functionname)"

**Comment by Moealfadil** [comment]
check the py_formating using ruff

File: solutions/tests/test_binary_to_hex.py (hunk 2 above)
Comment (MarcDarazi99) [reply solutions/tests/test_binary_to_hex.py:19]: done

File: solutions/binary_to_hex.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    Returns:
+        str: The hexadecimal representation of the binary number.
+
+    Raises:
+        ValueError: If the input is not a valid binary string.
+    
+    Example:
+        >>> binary_to_hex("1101")
+        'D'
+        >>> binary_to_hex("11110000")
```
[the remark this answers is not in the document]
Comment (MarcDarazi99) [reply solutions/binary_to_hex.py:35]: done

**Comment by MarcDarazi99** [comment]
> check the py_formating using ruff

how it could be solved ?

**Comment by Moealfadil** [comment]
download ruff in your vscode using "pip install ruff" or "pip3 install ruff"
format your needed files using "ruff format path/to/file"
you are suppose to receive an output (" file formatted")

**Comment by MarcDarazi99** [comment]
> download ruff in your vscode using "pip install ruff" or "pip3 install ruff" format your needed files using "ruff format path/to/file" you are suppose to receive an output (" file formatted")

how to install "pip install ruff" or "pip3 install ruff" ?

**Comment by Moealfadil** [comment]
write this in vs code terminal, make sure python is installed in your computer
you can also use ruff as a vscode extension but i have not used it before so im not sure how it works

**Review by HebaShaheen — COMMENTED** [review]
Delete the function from the test file.it should has only the class of unit tests.
